addons/web/models/ir_http.py

```diff
@@ -116,3 +118,65 @@ def get_currencies(self):
+        )
+        if status != 200:
+            return self._response_by_status(status, headers, content)
+        else:
+            content_base64 = base64.b64decode(content)
+            headers.append(('Content-Length', len(content_base64)))
+            response = request.make_response(content_base64, headers)
+        return response
+
+    @api.model
+    def _content_image(self, xmlid=None, model='ir.attachment', id=None, field='datas',
+            filename_field='name', unique=None, filename=None, mimetype=None, download=None,
+            width=0, height=0, crop=False, quality=0, access_token=None, **kwargs):
+        status, headers, image_base64 = self.binary_content(
+            xmlid=xmlid, model=model, id=id, field=field, unique=unique, filename=filename,
+            filename_field=filename_field, download=download, mimetype=mimetype,
+            default_mimetype='image/png', access_token=access_token
+        )
+        return self._content_image_get_response(
+            status, headers, image_base64, model=model, id=id, field=field, download=download,
+            width=width, height=height, crop=crop, quality=quality)
+
+    @api.model
+    def _content_image_get_response(self, status, headers, image_base64, model='ir.attachment',
+            id=None, field='datas', download=None, width=0, height=0, crop=False, quality=0):
+        if status in [301, 304] or (status != 200 and download):
+            return self._response_by_status(status, headers, image_base64)
+        if not image_base64:
+            placeholder_filename = False
+            if model in self.env:
+                placeholder_filename = self.env[model]._get_placeholder_filename(field)
+            placeholder_content = self._placeholder(image=placeholder_filename)
+            # Since we set a placeholder for any missing image, the status must be 200. In case one
+            # wants to configure a specific 404 page (e.g. though nginx), a 404 status will cause
+            # troubles.
+            status = 200
+            image_base64 = base64.b64encode(placeholder_content)
+            if not (width or height):
+                width, height = odoo.tools.image_guess_size_from_field_name(field)
+        try:
+            image_base64 = image_process(image_base64, size=(int(width), int(height)), crop=crop, quality=int(quality))
+        except Exception:
+            return request.not_found()
+        content = base64.b64decode(image_base64)
+        headers = http.set_safe_image_headers(headers, content)
+        response = request.make_response(content, headers)
+        response.status_code = status
+        return response
+
+    @api.model
+    def _placeholder(self, image=False):
+        if not image:
+            image = 'placeholder.png'
+        image_path = image.lstrip('/').split('/') if '/' in image else ['web', 'static', 'img', image]
+        with file_open(get_resource_path(*image_path), 'rb') as fd:
```

Comment:
No that I'm thinking about it:

* get_resource_path shoudl just be `os.path.join`, everything it's going is redundant with `file_open`
* in fact I'm not sure working with list is useful at all as `/` as separator should work on all platform we support
* more importantly, `file_open` now has a `filter_ext` which is a list of acceptable extensions. It should very much be used here, even if the current callers look safe (as all current implementations of `_get_placeholder_filename` return static paths) I don't think there's any reason to allow other files than, say, png and jpg?

Reply:
It was just a move in the first place, but I cleaned it as requested. Now just calling `file_open` with the string, and giving `filter_ext`.

Reply:
:+1: